### ISSUE TITLE
Access violation - WinRT exception serialization (WindowsAppSdk)

### ISSUE THREAD

**Issue by xpetrik** [MindscapeHQ/raygun4net#547]
**Describe the bug**
Crash occurs during the serialization of a WinRT exception using SimpleJson

System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'

**To Reproduce**
1. Demo project based on https://github.com/microsoft/WindowsAppSDK-Samples/tree/main/Samples/AppLifecycle/Restart
2. Project contains sources from your project with modification of visbility of method BuildMessage (public)
3. Folowing code crash on AccessViolatiaon (this is only the essential part of the method to send the report)
```csharp
var raygunClient = new RaygunClient("apikey");              
var message = await raygunClient.BuildMessage(ex, null);
string payload = SimpleJson.SerializeObject(message);
```

**Expected behavior**
no crash and send report

**Screenshots**
![image](https://github.com/user-attachments/assets/c2e0741b-a6f6-4bc8-a666-d985f9c033ca)

**Desktop:**
 - OS: Windows 11 23H2 x64
 - Visual Studio 2022

[cs-winui-packaged.zip](https://github.com/user-attachments/files/17349736/cs-winui-packaged.zip)

**Comment by QuantumNightmare** [comment]
Thanks for reporting this. Unfortunately I have not been able to get your repo project running so far.

I have not worked with WinUI before, so I first installed the necessary components for Visual Studio 2022. When attempting to run, it complained about the WindowsSdkPackageVersion and that the auto generated files contained unsafe code. I also had to delete the win-arm64 profile and the launchsettings file. After all this, it compiles and runs, but then encounters `System.Runtime.InteropServices.COMException: 'Class not registered (0x80040154 (REGDB_E_CLASSNOTREG))'`

So I then created a new WinUI packaged project which compiled and ran just fine. I copied over the Raygun4Net code from your project and the 3 lines of usage code that you provided, which resulted in correctly serializing the message.

Do you know the key differences between the repo you provided and a brand new WinUI packaged project? I have provided my project below in case it helps. The project is called WinUIErrorHarness. The solution contains your original repo along with the modifications I made in an attempt to get it running on my machine.

Is the screenshot you provided within the bowels of the serialization process? Do you know what object and property is trying to be accessed at that point?

[WinUIErrorHarness.zip](https://github.com/user-attachments/files/17357935/WinUIErrorHarness.zip)

**Comment by xpetrik** [comment]
Hi @QuantumNightmare
Your project also works, but NullReferenceException is not WinRT exception. 
https://en.wikipedia.org/wiki/Windows_Runtime

WinRT exception is some kind of COMException. This type of exception is common in WindowsAppSdk. MAUI is also based on WindowsAppSdk. My example contains this important code

```csharp
var pm = new PackageManager();
var result = await pm.AddPackageByAppInstallerFileAsync(new Uri("https://www.google.com"),
     AddPackageByAppInstallerOptions.ForceTargetAppShutdown, pm.GetDefaultPackageVolume());
```
This code expects the url where the xml file will be located (appinstaller) and tries to install a new version of the application. www.google.com does not contain an xml file and the WinRT layer will fail. There are of course countless ways to create a WinRT exception. 

https://[systemartrecordings.blob.core.windows.net/recordings/ewptN7Itmw.mp4](https://systemartrecordings.blob.core.windows.net/recordings/ewptN7Itmw.mp4)

This is probably the property that will cause the crash. This property has a getter and is evaluated later, but you have to do some research.
![image](https://github.com/user-attachments/assets/5282d9c9-4534-4c0e-8b9a-825eef67eb88)

David

**Comment by QuantumNightmare** [comment]
Hey David, Thanks for pointing this out. Simply putting that PackageManager code into the new project I created was enough to reproduce the issue.

The problem ultimately is that one of the objects in the Exception's Data dictionary (which gets included in the Raygun crash reporting message) is of type `WinRT.ObjectReferenceWithContext<WinRT.Interop.IUnknownVftbl>` which is having trouble being serialized.

A work around for this issue is to attach a `CustomGroupingKey` event hook to the RaygunClient:

`raygunClient.CustomGroupingKey += RaygunClientOnCustomGroupingKey;`

You could either clear the whole Data dictionary:

```
private void RaygunClientOnCustomGroupingKey(object sender, RaygunCustomGroupingKeyEventArgs e)
{
  e.Message?.Details?.Error?.Data?.Clear();
}
```

Or, if you need some of the other WinRT Exception Data dictionary values, or you've added your own, then just remove the problematic one:

```
private void RaygunClientOnCustomGroupingKey(object sender, RaygunCustomGroupingKeyEventArgs e)
{
  e.Message.Details.Error.Data.Remove("__RestrictedErrorObjectReference");
}
```

Note that in both examples, this is just modifying the Raygun message model instance and doesn't affect the original Exception object.

Here is a modified project including these examples:
[WinUIErrorHarness.zip](https://github.com/user-attachments/files/17369571/WinUIErrorHarness.zip)

We'll have a think about how best to include this in Raygun4Net so that it can just work.

Thanks again for bringing this issue to our attention and being very helpful with reproducing and understanding it. Let us know if you have further questions about this.

**Comment by xpetrik** [comment]
Hi @QuantumNightmare ,
we'll try a workaround, of course, and let you know. 

However, it would be useful to solve such a problem somehow in general in your code.  MAUI is one of your supported platforms. You don't write about WindowsAppSdk support directly, but it is a UWP replacement and according to Microsoft the future of desktop applications. Microsoft is actively rewriting their desktop applications (file explorer etc) under this framework.

The very unpleasant thing is that when you catch a harmless exception, your application crashes after try to send data. Your potential customer will definitely not be happy about this. He probably won't even want to read the FAQ section with workarounds :-)

Thanks so much for the response and we'll let you know.

**Comment by QuantumNightmare** [comment]
Absolutely! We're grateful that you took the time to bring it to our attention, even though it took a little back and forth to reproduce the issue 😄 I created a ticket for this a few days ago which will go through the prioritization process shortly - I added my thoughts about this being critical. It's a pain that it doesn't get stopped by the try/catch block, but we have a few ideas of how to avoid this issue.
